### PR TITLE
GH-424: Add MimeType JSON Deserializer

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-kafka/issues/424

Support `o.s.utils.MimeType` in headers.

__cherry-pick to 1.3.x__

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/support/DefaultKafkaHeaderMapper.java [hunk 1]
```diff
@@ -152,6 +156,9 @@ public DefaultKafkaHeaderMapper(ObjectMapper objectMapper, String... patterns) {
 		for (String pattern : patterns) {
 			this.matchers.add(new SimplePatternBasedHeaderMatcher(pattern));
 		}
+		@SuppressWarnings({ "rawtypes", "unchecked" })
+		Module module = new SimpleModule().addDeserializer(MimeType.class, new MimeTypeJsonDeserializer(objectMapper));
+		this.objectMapper.registerModule(module);
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/DefaultKafkaHeaderMapper.java]: That's cool of course!
But is it OK to mutate externally provided `ObjectMapper`? 
Especially in Spring Boot app when too much is based on globally auto-configured bean...

I won't mind though if we just document this with the IMPORTANT mark.

File: spring-kafka/src/main/java/org/springframework/kafka/support/MimeTypeJsonDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * {@link MimeType} deserializer.
+ *
+ * @param <T> the type.
+ *
+ * @author Gary Russell
+ * @since 1.3
+ *
+ */
+public class MimeTypeJsonDeserializer<T extends MimeType> extends StdNodeBasedDeserializer<T> {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/MimeTypeJsonDeserializer.java]: Not sure what made you to have this strange generics configuration, but simplified version works well for me:
https://github.com/artembilan/spring-kafka/commit/5511ad52ab530c40d46fbdc5c289ce6b6d798a67

Am I missing anything?
Thanks

Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/support/MimeTypeJsonDeserializer.java]: I had that originally but had some problem - must have been pilot error; feel free to use your version.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/398b107dd9cca54ff807a3fddf3dde5915216aaf and cherry-picked to `1.3.x`

**Pull request closed without merging**